### PR TITLE
Calendar tap and hold

### LINKED ISSUE #2872: Enable tap and drag to create time entries in the calendar collection view
Depends on #2871

We need to track the hold/drag gestures on empty areas of the collection view (this was already considered from ground up, so it should already be working) and bind these events to the action added in #2871. This way we can create time entries from the calendar view.

### LINKED ISSUE #2965: The current time indicator is always in GMT+0
I'm Brazilian, not British. Timezones are hard. Let's fix this!

### PR DESCRIPTION
**Pull request by juanlaube**
## What's this?

Adds the tap and hold functionality to create a time entry form a timespan.

### Relationships

Closes #2872
Accidentally closes #2965 

## How is it done?

I wrapped a long press gesture recognized in a helper class, from there it's all calculations based on the gesture point and state.

### Side effects

None.

## Review hints

Please test, test, test and test. Let me know of any bug, we can fix some in this pull request or decide to split that into a new issue.

## :squid: Permissions

Myself.

### CODE REVIEW

File: Toggl.Daneel/ViewControllers/Calendar/CalendarViewController.cs [hunk 1]
```diff
@@ -33,13 +34,17 @@ public override void ViewDidLoad()
...
             layout = new CalendarCollectionViewLayout(timeService, dataSource);
 
+            longPressHelper = new CalendarCollectionViewLongPressHelper(CalendarCollectionView, dataSource, layout);
+
             CalendarCollectionView.SetCollectionViewLayout(layout, false);
             CalendarCollectionView.Delegate = dataSource;
             CalendarCollectionView.DataSource = dataSource;
 
             this.Bind(dataSource.ItemTapped, ViewModel.OnItemTapped);
+            this.Bind(dataSource.CreateFromSpan, ViewModel.OnDurationSelected);
```
Comment (willsb) [inline Toggl.Daneel/ViewControllers/Calendar/CalendarViewController.cs]: Can you explain the reasoning behind exposing this in the dataSource and not in the long press helper?

File: Toggl.Daneel/ViewSources/CalendarCollectionViewSource.cs [hunk 2]
```diff
@@ -50,6 +52,7 @@ public sealed class CalendarCollectionViewSource : MvxCollectionViewSource, ICal
             Ensure.Argument.IsNotNull(date, nameof(date));
             Ensure.Argument.IsNotNull(timeOfDayFormat, nameof(timeOfDayFormat));
             Ensure.Argument.IsNotNull(collection, nameof(collection));
+            this.date = date;
```
Comment (willsb) [inline Toggl.Daneel/ViewSources/CalendarCollectionViewSource.cs]: who dis? 👀

File: Toggl.Daneel/Views/Calendar/CalendarCollectionViewLayout.cs [hunk 3]
```diff
@@ -148,7 +160,7 @@ private CGRect frameForHour(int hour)
 
         private CGRect frameForCurrentTime()
         {
-            var now = timeService.CurrentDateTime;
+            var now = timeService.CurrentDateTime.LocalDateTime;
```
Comment (willsb) [inline Toggl.Daneel/Views/Calendar/CalendarCollectionViewLayout.cs]: I assume this is what collaterally fixes the other issue, correct?

File: Toggl.Daneel/Views/Calendar/CalendarCollectionViewLongPressHelper.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+        private UILongPressGestureRecognizer longPressGestureRecognizer;
+
+        private NSIndexPath itemIndexPath;
+
+        private DateTimeOffset timeAtFirstPoint;
+        private CGPoint? firstPoint;
+        private CGPoint? lastPoint;
+
+        private TimeSpan defaultDuration = TimeSpan.FromMinutes(15);
```
Comment (willsb) [inline Toggl.Daneel/Views/Calendar/CalendarCollectionViewLongPressHelper.cs]: `private static readonly`

File: Toggl.Daneel/Views/Calendar/CalendarCollectionViewLongPressHelper.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                    itemIndexPath = null;
+                    break;
+
+                case UIGestureRecognizerState.Cancelled:
+                case UIGestureRecognizerState.Failed:
+                    dataSource.RollbackChanges();
+                    itemIndexPath = null;
+                    break;
+
+                default:
```
Comment (willsb) [inline Toggl.Daneel/Views/Calendar/CalendarCollectionViewLongPressHelper.cs]: These lines can be removed

File: Toggl.Daneel/Views/Calendar/CalendarCollectionViewLongPressHelper.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            {
+                case UIGestureRecognizerState.Began:
+                    timeAtFirstPoint = layout.DateAtPoint(point);
+                    itemIndexPath = dataSource.InsertItem(timeAtFirstPoint, defaultDuration);
+                    firstPoint = point;
+                    lastPoint = point;
+                    break;
+
+                case UIGestureRecognizerState.Changed:
+                    if (firstPoint.HasValue && firstPoint.Value.Y >= point.Y)
```
Comment (willsb) [inline Toggl.Daneel/Views/Calendar/CalendarCollectionViewLongPressHelper.cs]: Why not allow the user to scroll up, making the anchor point the effective end of the time entry?

File: Toggl.Daneel/ViewControllers/Calendar/CalendarViewController.cs (hunk 1 above)
Comment (juanlaube) [reply Toggl.Daneel/ViewControllers/Calendar/CalendarViewController.cs]: For some reason it made sense first, but it doesn't, I will try to move this to the long press helper thing.

File: Toggl.Daneel/ViewSources/CalendarCollectionViewSource.cs (hunk 2 above)
Comment (juanlaube) [reply Toggl.Daneel/ViewSources/CalendarCollectionViewSource.cs]: This removes a warning for an unassigned field. I'd like to improve this later, all the pieces of the collection view (data source, layout, long press) depends on the date for some calculations. I'd like to improve how we're passing that date, but that's for another PR.

File: Toggl.Daneel/Views/Calendar/CalendarCollectionViewLayout.cs (hunk 3 above)
Comment (juanlaube) [reply Toggl.Daneel/Views/Calendar/CalendarCollectionViewLayout.cs]: Correct 🙂

File: Toggl.Daneel/Views/Calendar/CalendarCollectionViewLongPressHelper.cs (hunk 6 above)
Comment (juanlaube) [reply Toggl.Daneel/Views/Calendar/CalendarCollectionViewLongPressHelper.cs]: Should be easy to do 🙂

File: Toggl.Foundation/Calendar/CalendarItem.cs [hunk 7]
```diff
@@ -51,5 +51,25 @@ public CalendarItem(IThreadSafeTimeEntry timeEntry)
...
+                this.Source,
+                startTime,
+                this.Duration,
+                this.Description,
+                this.Color,
+                this.TimeEntryId,
+                this.CalendarId);
+
+        public CalendarItem WithDuration(TimeSpan duration)
+        => new CalendarItem(
```
[the remark this answers is not in the document]
Comment (juanlaube) [reply Toggl.Foundation/Calendar/CalendarItem.cs]: Thanks VS 🙂

**Pull request merged**